xml_converter/src/xml_converter.cpp

```diff
@@ -61,30 +63,115 @@ void write_xml_file(string xml_filepath, map<string, Category>* marker_categorie
+                    }
+                }
+            }
+            ++keep;
+        }
+    }
+
+    proto_category->mutable_children()->DeleteSubrange(keep, proto_category->children_size() - keep);
+}
+
 void write_protobuf_file(string proto_filepath, map<string, Category>* marker_categories, vector<Parseable*>* parsed_pois) {
-    ofstream outfile;
-    waypoint::Waypoint proto_message;
+    waypoint::Waypoint proto_pois;
+    waypoint::Waypoint all_categories;
+    waypoint::Waypoint output_message;
+    std::set<int> map_ids;
+    std::set<string> proto_categories;
+    std::set<string> category_includes;
+
+    if (mkdir(proto_filepath.c_str(), 0700) == -1) {
+        cout << "Error making ./protobins" << endl;
+        throw std::error_code();
+    }
 
-    outfile.open(proto_filepath, ios::out | ios::binary);
     for (const auto& category : *marker_categories) {
         waypoint::Category proto_category = category.second.as_protobuf();
-        proto_message.add_category()->CopyFrom(proto_category);
+        all_categories.add_category()->CopyFrom(proto_category);
+        create_proto_categories(proto_category, &proto_categories);
     }
 
     for (const auto& parsed_poi : *parsed_pois) {
         if (parsed_poi->classname() == "POI") {
             Icon* icon = dynamic_cast<Icon*>(parsed_poi);
             waypoint::Icon poi = icon->as_protobuf();
-            proto_message.add_icon()->CopyFrom(poi);
+            proto_pois.add_icon()->CopyFrom(poi);
+            map_ids.insert(icon->map_id);
         }
         if (parsed_poi->classname() == "Trail") {
             Trail* trail = dynamic_cast<Trail*>(parsed_poi);
             waypoint::Trail poi = trail->as_protobuf();
-            proto_message.add_trail()->CopyFrom(poi);
+            proto_pois.add_trail()->CopyFrom(poi);
+            map_ids.insert(trail->map_id);
         }
     }
-    proto_message.SerializeToOstream(&outfile);
-    outfile.close();
+
+    for (int map_id : map_ids) {
+        ofstream outfile;
+        string output_filepath = proto_filepath + "/" + to_string(map_id) + ".data";
+        outfile.open(output_filepath, ios::out | ios::binary);
+        output_message.CopyFrom(all_categories);
+
+        for (int i = 0; i < proto_pois.icon_size(); i++) {
+            vector<string> split_categories = split(proto_pois.icon(i).category().name(), ".");
+            if (proto_pois.icon(i).map_id() == map_id) {
+                auto category = proto_categories.find(split_categories[split_categories.size() - 1]);
+                if (category == proto_categories.end()) {
+                    cout << "Unknown MarkerCategory " << proto_pois.icon(i).category().name() << endl;
+                }
+                else {
+                    for (string split_category : split_categories) {
+                        category_includes.insert(split_category);
+                    }
+                }
+                output_message.add_icon()->CopyFrom(proto_pois.icon(i));
```

Comment:
In a future step we will need to think about how to handle this case visually. How does one hide/show something without a category. We might want to create missing categories so that they can be shown and hidden. This should probably not happen here though.

It brings up the interesting point of: if every POI or Trail has a category, should the cateogry proto just own them instead of having a separate array. This would change the proto structure pretty significantly.